File: sdks/ruby/examples/typed_greet_node.rb
# frozen_string_literal: true

require_relative "../lib/blok"

# Typed greeting node demonstrating the SPEC-B TypedNode contract.
class TypedGreetNode < Blok::Node::TypedNode
  node_name "typed-greet"
  description "Typed greeting (SPEC-B contract demo)"

  input do
    field :name, :string, required: true
    field :repeat, :integer, default: 1
  end

  output do
    field :greeting, :string
    field :length, :integer
  end

  def run(_ctx, input)
    repeat = input[:repeat].is_a?(Integer) && input[:repeat].positive? ? input[:repeat] : 1
    greeting = ("Hello, " + input[:name]) * repeat
    { "greeting" => greeting, "length" => greeting.length }
  end
end

File: sdks/elixir/examples/typed_greet_node.ex
defmodule Blok.Examples.TypedGreetInput do
  use Blok.Schema
  field :name, :string, required: true
end

defmodule Blok.Examples.TypedGreetOutput do
  use Blok.Schema
  field :message, :string, required: true
end

defmodule Blok.Examples.TypedGreet do
  use Blok.Node,
    name: "typed-greet",
    description: "Returns a greeting from the supervised BEAM runtime",
    input: Blok.Examples.TypedGreetInput,
    output: Blok.Examples.TypedGreetOutput,
    capability_manifest: %{"version" => "1", "classification" => "agent-compatible", "secret_refs" => []}

  @impl Blok.Node
  def execute(_ctx, input), do: {:ok, %Blok.Examples.TypedGreetOutput{message: "Hello, #{input.name}!"}}
end
